org.mwc.debrief.legacy/src/Debrief/Tools/Palette/CreateLabel.java

```diff
@@ -133,28 +136,73 @@ public final Action getData()
       final LabelWrapper theWrapper = new LabelWrapper("blank label",
           centre,
           MWC.GUI.Properties.DebriefColors.ORANGE);
+      Layer theLayer = null;
+      String layerToAddTo = getSelectedLayer();
+      if(CoreCreateShape.USER_SELECTED_LAYER_COMMAND.equals(layerToAddTo) || Layers.NEW_LAYER_COMMAND.equals(layerToAddTo)) {
+        userSelected=true;
+        if(CoreCreateShape.USER_SELECTED_LAYER_COMMAND.equals(layerToAddTo)) {
+          layerToAddTo = getLayerName();
+        }
+        else {
+          String txt = JOptionPane.showInputDialog(null, "Enter name for new layer");
+          // check there's something there
+          if (txt!=null && !txt.isEmpty())
+          {
+            layerToAddTo = txt;
+            // create base layer
+            final Layer newLayer = new BaseLayer();
+            newLayer.setName(layerToAddTo);
+
+            // add to layers object
+            _theData.addThisLayer(newLayer);
+          }
+        }
+        if(layerToAddTo!=null) {
+          theLayer = _theData.findLayer(layerToAddTo);
+        }

```

Comment:
@ayeshare - the logic is confused in this method.  It's not necessarily down to you, but I'd appreciate you going through this refactoring exercise.

We're duplicating the `layer != null` test on the next two lines. Then we create a `res` value that I think gets over-written later on.

I know it's a chore, but could you re-work the logic in the whole of this method, to verify control flows as expected.  

Again, I know it's a chore, but I suggest you declare res as `final Action res;` - this will ensure it both gets assigned _somewhere_, and that it only gets assigned once.